File: main/src/unity/world/blocks/Teleporter.java
package unity.world.blocks;

import arc.Events;
import arc.graphics.Color;
import arc.graphics.g2d.*;
import arc.math.Mathf;
import arc.scene.ui.*;
import arc.scene.ui.layout.Table;
import arc.struct.*;
import arc.util.*;
import arc.util.io.*;
import mindustry.entities.units.BuildPlan;
import mindustry.game.Team;
import mindustry.game.EventType.WorldLoadEvent;
import mindustry.gen.*;
import mindustry.type.Item;
import mindustry.ui.Styles;
import mindustry.world.Block;

import static arc.Core.*;

public class Teleporter extends Block{
    protected static final Color[] selection = new Color[]{Color.royal, Color.orange, Color.scarlet, Color.forest, Color.purple, Color.gold, Color.pink, Color.black};
    protected static final ObjectSet<TeleporterBuild>[][] teleporters;
    protected float powerUse = 2.5f;
    protected TextureRegion blankRegion, topRegion;

    static{
        teleporters = new ObjectSet[Team.baseTeams.length][selection.length];
        for(int i = 0; i < Team.baseTeams.length; i++){
            if(teleporters[i] == null) teleporters[i] = new ObjectSet[selection.length];
            for(int j = 0; j < selection.length; j++) teleporters[i][j] = new ObjectSet<>();
        }
    }

    public Teleporter(String name){
        super(name);
        update = true;
        solid = true;
        configurable = true;
        saveConfig = true;
        unloadable = false;
        hasItems = true;
        Events.on(WorldLoadEvent.class, e -> {
            for(int i = 0; i < teleporters.length; i++){
                for(int j = 0; j < teleporters[i].length; j++) teleporters[i][j].clear();
            }
        });
        config(Integer.class, (TeleporterBuild build, Integer value) -> {
            if(build.toggle != -1) teleporters[build.team.id][build.toggle].remove(build);
            if(value != -1) teleporters[build.team.id][value].add(build);
            build.toggle = value;
        });
        configClear((TeleporterBuild build) -> build.toggle = -1);
    }

    @Override
    public boolean outputsItems(){
        return true;
    }

    @Override
    public void init(){
        consumes.powerCond(powerUse, TeleporterBuild::isConsuming);
        super.init();
    }

    @Override
    public void load(){
        super.load();
        blankRegion = atlas.find(name + "-blank");
        topRegion = atlas.find(name + "-top");
    }

    @Override
    public void drawRequestConfig(BuildPlan req, Eachable<BuildPlan> list){
        drawRequestConfigCenter(req, req.config, "nothing");
    }

    @Override
    public void drawRequestConfigCenter(BuildPlan req, Object content, String region){
        if(!(content instanceof Integer)) return;
        int temp = (int) content;
        Draw.color(selection[temp]);
        Draw.rect(blankRegion, req.drawx(), req.drawy());
    }

    public class TeleporterBuild extends Building{
        protected int toggle = -1, entry;
        protected float duration;
        protected TeleporterBuild target;
        protected Team previousTeam;

        protected void onDuration(){
            if(duration < 0f) duration = 0f;
            else duration -= Time.delta;
        }

        protected boolean isConsuming(){
            return duration > 0f;
        }

        protected boolean isTeamChanged(){
            return previousTeam != team;
        }

        @Override
        public void draw(){
            super.draw();
            if(toggle != -1){
                Draw.color(selection[toggle]);
                Draw.rect(blankRegion, x, y);
            }
            Draw.color(Color.white);
            Draw.alpha(0.45f + Mathf.absin(Time.time(), 7f, 0.26f));
            Draw.rect(topRegion, x, y);
            Draw.reset();
        }

        @Override
        public void updateTile(){
            onDuration();
            if(items.any()) dump();
            if(isTeamChanged() && toggle != -1){
                teleporters[team.id][toggle].add(this);
                teleporters[previousTeam.id][toggle].remove(this);
                previousTeam = team;
            }
        }

        @Override
        public void buildConfiguration(Table table){
            final ButtonGroup<Button> group = new ButtonGroup<>();
            group.setMinCheckCount(0);
            for(int i = 0; i < selection.length; i++){
                int j = i;
                ImageButton button = table.button(Tex.whiteui, Styles.clearToggleTransi, 24f, () -> {}).size(34f).group(group).get();
                button.changed(() -> configure(button.isChecked() ? j : -1));
                button.getStyle().imageUpColor = selection[j];
                button.update(() -> button.setChecked(toggle == j));
                if(i % 4 == 3) table.row();
            }
        }

        protected TeleporterBuild findLink(int value){
            ObjectSet<TeleporterBuild> teles = teleporters[team.id][value];
            Seq<TeleporterBuild> entries = teles.asArray();
            if(entry >= entries.size) entry = 0;
            if(entry == entries.size - 1){
                TeleporterBuild other = teles.get(entries.get(entry));
                if(other == this) entry = 0;
            }
            for(int i = entry, len = entries.size; i < len; i++){
                TeleporterBuild other = teles.get(entries.get(i));
                if(other != this){
                    entry = i + 1;
                    return other;
                }
            }
            return null;
        }

        @Override
        public boolean acceptItem(Building source, Item item){
            if(toggle == -1) return false;
            target = findLink(toggle);
            if(target == null) return false;
            return source != this && cons.valid() && Mathf.zero(1 - efficiency()) && target.items.total() < target.getMaximumAccepted(item);
        }

        @Override
        public void handleItem(Building source, Item item){
            target.items.add(item, 1);
            duration = 0f;
        }

        @Override
        public void created(){
            if(toggle != -1) teleporters[team.id][toggle].add(this);
            previousTeam = team;
        }

        @Override
        public void onRemoved(){
            if(toggle != -1){
                if(isTeamChanged()) teleporters[previousTeam.id][toggle].remove(this);
                else teleporters[team.id][toggle].remove(this);
            }
            //unity.Unity.print(teleporters[team.id]);
        }

        @Override
        public Integer config(){
            return toggle;
        }

        @Override
        public void write(Writes write){
            super.write(write);
            write.b(toggle);
        }

        @Override
        public void read(Reads read, byte revision){
            super.read(read, revision);
            toggle = read.b();
        }
    }
}

File: scripts/koruh/blocks/teleporter.js
const selection = [Color.royal, Color.orange, Color.scarlet, Color.forest, Color.purple, Color.gold, Color.pink, Color.black];
const teleporters = [];
Events.on(WorldLoadEvent, e => {
    //print("claerTeleporters");
    for(var i in Team.baseTeams) {
        var tmp = Team.get(i).toString();
        if(teleporters[tmp] == null) teleporters[tmp] = [];
        for(var j = 0; j < selection.length; j++) {
            if(teleporters[tmp][j] != null) teleporters[tmp][j].clear();
            else teleporters[tmp].push(new ObjectSet());
        }
    }
});
const teleporter = extendContent(Block, "teleporter", {
    lastColor: -1,
    /*playerPlaced(tile) {
        if (this.lastColor != null) tile.configure(this.lastColor);
    },*/
    outputsItems() {
        return true;
    },
    init() {
        this.hasPower = true;
        this.consumes.powerCond(2.5, boolf(entity => entity != null ? entity.isConsuming() : false));
        this.super$init();
    },
    load() {
        this.super$load();
        this.blankRegion = Core.atlas.find(this.name + "-blank");
        this.topRegion = Core.atlas.find(this.name + "-top");
    },
    drawRequestConfig(req, list) {
        this.drawRequestConfigCenter(req, req.config, "blank");
    },
    drawRequestConfigCenter(req, content, region) {
        if(isNaN(content) || selection[content] == null) return;
        Draw.color(selection[content]);
        Draw.rect(this.blankRegion, req.drawx(), req.drawy());
        Draw.color();
    }
});
teleporter.update = true;
teleporter.solid = true;
teleporter.configurable = true;
teleporter.saveConfig = true;
teleporter.unloadable = false;
teleporter.buildType = prov(() => extend(Building, {
    draw() {
        this.super$draw();
        if(this.toggle != -1) {
            Draw.color(selection[this.toggle]);
            Draw.rect(teleporter.blankRegion, this.x, this.y);
        }
        Draw.color(Color.white);
        Draw.alpha(0.45 + Mathf.absin(Time.time(), 7, 0.26));
        Draw.rect(teleporter.topRegion, this.x, this.y);
        Draw.reset();
    },
    updateTile() {
        this.onDuration();
        if(this.items.total() > 0) this.dump();
        if(this.isTeamChanged() && this.toggle != -1) {
            teleporters[this.team.toString()][this.toggle].add(this);
            teleporters[this.previousTeam.toString()][this.toggle].remove(this);
            this.previousTeam = this.team;
        }
    },
    buildConfiguration(table) {
        const cont = new Table();
        const group = new ButtonGroup();
        group.setMinCheckCount(0);
        for(var i = 0; i < selection.length; i++) {
            (function(i, entity) {
                var button = cont.button(Tex.whiteui, Styles.clearToggleTransi, 24, run(() => {
                    teleporter.lastColor = button.isChecked() ? i : -1;
                    entity.configure(button.isChecked() ? i : -1);
                })).size(34, 34).group(group).get();
                button.getStyle().imageUpColor = selection[i];
                button.setChecked(entity != null ? entity.toggle == i ? true : false : false);
            })(i, this)
            if(i % 4 == 3) cont.row();
        }
        table.add(cont);
    },
    configured(player, value) {
        if(isNaN(value) || selection[value] == null) return;
        var team = this.team.toString();
        if(this.toggle != -1) teleporters[team][this.toggle].remove(this);
        if(value != -1) teleporters[team][value].add(this);
        this.toggle = value;
    },
    findLink(value) {
        var valueTeles = teleporters[this.team.toString()][value];
        var entries = valueTeles.asArray();
        if(this.entry >= entries.size) this.resetEntry();
        if(this.entry == entries.size - 1) {
            var other_ = valueTeles.get(entries.get(this.entry))
            if(other_ == this) this.resetEntry();
        }
        for(var i = this.entry; i < entries.size; i++) {
            var other = valueTeles.get(entries.get(i))
            if(other != this) {
                this.entry = i + 1;
                return other;
            }
        }
        return null;
    },
    acceptItem(source, item) {
        if(this.toggle == -1) return false;
        this.target = this.findLink(this.toggle);
        if(this.target == null) return false;
        return source != this && this.cons.valid() && Mathf.zero(1 - this.efficiency()) && this.target.items.total() < this.target.getMaximumAccepted(item);
    },
    handleItem(source, item) {
        this.target.items.add(item, 1);
        this.resetDuration();
    },
    toggle: -1,
    resetEntry() {
        this.entry = 0;
    },
    entry: 0,
    target: null,
    resetDuration() {
        this.duration = 60;
    },
    onDuration() {
        if(this.duration <= 0) this.duration = 0;
        else this.duration -= Time.delta;
    },
    isConsuming() {
        return this.duration > 0;
    },
    duration: 0,
    previousTeam: null,
    isTeamChanged() {
        return this.previousTeam != this.team;
    },
    created() {
        if(this.toggle != -1) teleporters[this.team.toString()][this.toggle].add(this);
        this.previousTeam = this.team;
    },
    onRemoved() {
        //print(teleporters[this.team.toString()]);
        if(this.toggle != -1) {
            if(this.isTeamChanged()) teleporters[this.previousTeam.toString()][this.toggle].remove(this);
            else teleporters[this.team.toString()][this.toggle].remove(this);
        }
    },
    config() {
        return this.toggle;
    },
    write(writes) {
        this.super$write(writes);
        writes.b(this.toggle);
    },
    read(reads, revision) {
        this.super$read(reads, revision);
        this.toggle = reads.b();
    }
}));
